server/coordinator/procedure/ddl/droptable/drop_table.go

```diff
@@ -126,67 +127,117 @@ func prepareCallback(event *fsm.Event) {
+}
 
-	onSucceeded func(cluster.TableInfo) error
-	onFailed    func(error) error
+type ProcedureParams struct {
+	ID              uint64
+	Dispatch        eventdispatch.Dispatch
+	ClusterMetadata *metadata.ClusterMetadata
+	ClusterSnapshot metadata.Snapshot
 
-	ret cluster.TableInfo
+	SourceReq   *metaservicepb.DropTableRequest
+	OnSucceeded func(metadata.TableInfo) error
+	OnFailed    func(error) error
 }
 
-func NewDropTableProcedure(dispatch eventdispatch.Dispatch, cluster *cluster.Cluster, id uint64, req *metaservicepb.DropTableRequest, onSucceeded func(cluster.TableInfo) error, onFailed func(error) error) procedure.Procedure {
+func NewDropTableProcedure(params ProcedureParams) (procedure.Procedure, error) {
+	shardID, err := validateTable(params)
+	if err != nil {
+		return nil, err
+	}
+
+	relatedVersionInfo := buildRelatedVersionInfo(params, shardID)
+
 	fsm := fsm.NewFSM(
 		stateBegin,
 		dropTableEvents,
 		dropTableCallbacks,
 	)
+
 	return &Procedure{
-		id:          id,
-		fsm:         fsm,
-		cluster:     cluster,
-		dispatch:    dispatch,
-		req:         req,
-		onSucceeded: onSucceeded,
-		onFailed:    onFailed,
-		state:       procedure.StateInit,
+		fsm:                fsm,
+		shardID:            shardID,
+		relatedVersionInfo: relatedVersionInfo,
+		params:             params,
+		state:              procedure.StateInit,
+	}, nil
+}
+
+func buildRelatedVersionInfo(params ProcedureParams, shardID storage.ShardID) procedure.RelatedVersionInfo {
+	shardWithVersion := make(map[storage.ShardID]uint64, 1)
+	for _, shardView := range params.ClusterSnapshot.Topology.ShardViews {
+		if shardView.ShardID == shardID {
+			shardWithVersion[shardID] = shardView.Version
+		}
+	}
+	return procedure.RelatedVersionInfo{
+		ClusterID:        params.ClusterSnapshot.Topology.ClusterView.ClusterID,
+		ShardWithVersion: shardWithVersion,
+		ClusterVersion:   params.ClusterSnapshot.Topology.ClusterView.Version,
 	}
 }
 
-type Procedure struct {
-	id       uint64
-	fsm      *fsm.FSM
-	cluster  *cluster.Cluster
-	dispatch eventdispatch.Dispatch
-	req      *metaservicepb.DropTableRequest
+func validateTable(params ProcedureParams) (storage.ShardID, error) {
+	table, exists, err := params.ClusterMetadata.GetTable(params.SourceReq.GetSchemaName(), params.SourceReq.GetName())
+	if err != nil {
+		log.Error("get table", zap.Error(err))
+		return 0, err
+	}
+	if !exists {
+		log.Error("drop non-existing table", zap.String("schema", params.SourceReq.GetSchemaName()), zap.String("table", params.SourceReq.GetName()))
+		return 0, err
+	}
+
+	for _, shardView := range params.ClusterSnapshot.Topology.ShardViews {
+		for _, tableID := range shardView.TableIDs {
+			if table.ID == tableID {
+				return shardView.ShardID, nil
+			}
+		}
+	}
+
+	return 0, errors.WithMessage(metadata.ErrShardNotFound, fmt.Sprintf("The shard corresponding to the table was not found, schemaName:%s,tableName:%s", params.SourceReq.GetSchemaName(), params.SourceReq.GetName()))
```

Comment:
- Do we have `errors.WithMessagef`?
- `schemaName:%s,tableName:%s"` => `schema:%s, table:%s"`